### PR TITLE
Feature/queues

### PR DESCRIPTION
**Pull request by RexTremendae**
Please include a summary of the change. Please also include relevant motivation and context (if applicable).

- [X] I have performed a self-review of my code
- [X] My code follows the style guidelines of this project
- [X] I have added comments to hard-to-understand areas of my code
- [X] I have added tests that prove my fix is effective or that my feature works
- [X] New and existing unit tests pass locally with my changes
- [X] I have made corresponding changes to the comments
- [X] I have made corresponding changes to the README.md

### CODE REVIEW

File: DataStructures.Tests/ListBasedQueueTests.cs [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+            var q = new ListBasedQueue<char>();
+            q.Enqueue('A');
+            q.Enqueue('B');
+            q.Enqueue('C');
+
+            // Act
+            string result = "";
+            for (int i = 0; i < 3; i++)
+            {
+                result += q.Dequeue();
```
Comment (siriak) [inline DataStructures.Tests/ListBasedQueueTests.cs:21]: Codacy found an issue: [Use a StringBuilder instead.](https://app.codacy.com/manual/siriak/C-Sharp/pullRequest?prid=6271167)

**Comment by siriak** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 3
           

Complexity increasing per file
==============================
- DataStructures/ListBasedQueue.cs  2
- DataStructures.Tests/ListBasedQueueTests.cs  2
- DataStructures/ArrayBasedQueue.cs  3
- DataStructures.Tests/ArrayBasedQueueTests.cs  2
         

Clones added
============
- DataStructures.Tests/ListBasedQueueTests.cs  4
- DataStructures.Tests/ArrayBasedQueueTests.cs  4
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/siriak/C-Sharp/pullRequest?prid=6271167&bid=20414170)

File: DataStructures/ArrayBasedQueue.cs [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+    /// </summary>
+    /// <typeparam name="T">Generic Type.</typeparam>
+    public class ArrayBasedQueue<T>
+    {
+        /// <summary>
+        /// <see cref="List{T}"/> based queue.
+        /// </summary>
+        private readonly T[] queue;
+        private int currentIndex;
+        private int capacity;
```
Comment (siriak) [inline DataStructures/ArrayBasedQueue.cs:17]: Codacy found an issue: [Make 'capacity' 'readonly'.](https://app.codacy.com/manual/siriak/C-Sharp/pullRequest?prid=6271167)

File: DataStructures.Tests/ArrayBasedQueueTests.cs [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+            var q = new ArrayBasedQueue<char>(3);
+            q.Enqueue('A');
+            q.Enqueue('B');
+            q.Enqueue('C');
+
+            // Act
+            string result = "";
+            for (int i = 0; i < 3; i++)
+            {
+                result += q.Dequeue();
```
Comment (siriak) [inline DataStructures.Tests/ArrayBasedQueueTests.cs:21]: Codacy found an issue: [Use a StringBuilder instead.](https://app.codacy.com/manual/siriak/C-Sharp/pullRequest?prid=6271167)

**Comment by siriak** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- DataStructures/ListBasedQueue.cs  2
- DataStructures.Tests/ListBasedQueueTests.cs  2
- DataStructures/ArrayBasedQueue.cs  3
- DataStructures.Tests/ArrayBasedQueueTests.cs  2
         

Clones added
============
- DataStructures.Tests/ListBasedQueueTests.cs  4
- DataStructures.Tests/ArrayBasedQueueTests.cs  4
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/siriak/C-Sharp/pullRequest?prid=6271167&bid=20414170)

File: DataStructures/ArrayBasedQueue.cs [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+            if (IsEmpty())
+            {
+                throw new InvalidOperationException("There are no items in the queue.");
+            }
+
+            var item = queue[0];
+
+            for (int i = 0; i < currentIndex - 1; i++)
+            {
+                queue[i] = queue[i + 1];
```
Comment (siriak) [inline DataStructures/ArrayBasedQueue.cs:48]: Unfortunately, this is not the way to go. In queue Enqueue and Dequeue operations must be performed in O(1) time, but this is O(n). See [this](https://www.geeksforgeeks.org/circular-queue-set-1-introduction-array-implementation/) and [this](https://en.wikipedia.org/wiki/Circular_buffer) for some hints

**Review by siriak — CHANGES_REQUESTED** [review]
I've left a comment about complexity, everything else looks good, thanks!

File: DataStructures/ArrayBasedQueue.cs (hunk 4 above)
Comment (RexTremendae) [reply DataStructures/ArrayBasedQueue.cs]: Yeah, sorry for being lazy 😄

**Comment by siriak** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- DataStructures/ListBasedQueue.cs  2
- DataStructures.Tests/ListBasedQueueTests.cs  2
- DataStructures/ArrayBasedQueue.cs  3
- DataStructures.Tests/ArrayBasedQueueTests.cs  2
         

Clones added
============
- DataStructures.Tests/ListBasedQueueTests.cs  4
- DataStructures.Tests/ArrayBasedQueueTests.cs  4
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/siriak/C-Sharp/pullRequest?prid=6271167&bid=20414170)

**Review by siriak — APPROVED** [review]
You forgot to add queue types to README.md, please add them to the list and I'll merge your changes!

**Comment by siriak** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- DataStructures/ListBasedQueue.cs  2
- DataStructures.Tests/ListBasedQueueTests.cs  2
- DataStructures.Tests/ArrayBasedQueueTests.cs  2
- DataStructures/ArrayBasedQueue.cs  3
         

Clones added
============
- DataStructures.Tests/ListBasedQueueTests.cs  4
- DataStructures.Tests/ArrayBasedQueueTests.cs  4
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/siriak/C-Sharp/pullRequest?prid=6271167&bid=20414170)

**Review by siriak — APPROVED** [review]

**Pull request merged**